### ISSUE TITLE
space-before-function-paren doesn't detect async keyword

### ISSUE THREAD

**Issue by Redsandro** [eslint/eslint#10370]
**Tell us about your environment**

* **ESLint Version:** 4.18.1
* **Node Version:** 8.9.3
* **npm Version:** 3.10.10

**What parser (default, Babel-ESLint, etc.) are you using?**

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
{
    parserOptions: {
        ecmaVersion: 2017
    },
    "env": {
        "node": true,
        "browser": true,
        "es6": true
    },
    "extends": "eslint:recommended",
    "rules": {
        "indent": [
            1,
            "tab",
            { "SwitchCase": 1, "MemberExpression": "off" }
        ],
        "space-before-blocks": [
            1
        ],
        "space-before-function-paren": [
            1,
            "never"
        ],
        "quotes": [
            1,
            "single",
            "avoid-escape"
        ],
        "no-extra-parens" : [
            1,
            "all",
            { "returnAssign": false }
        ],
        "strict": [
            2,
            "safe"
        ],
        "brace-style": [
            1,
            "stroustrup",
            { "allowSingleLine": true }
        ],
        "no-var": [
            1
        ],
        "no-unused-vars": [
            1,
            {
                "args": "none"
            }
        ],
        "prefer-const": [
            1,
            {
                "destructuring": "any",
                "ignoreReadBeforeAssign": false
            }
        ],
        "no-multiple-empty-lines": [
            1,
            {
                "max": 1,
                "maxEOF": 1
            }
        ],
        "no-fallthrough": [
            1
        ],
        "no-octal": [
            1
        ],
        "no-console": [
            0
        ],
        "linebreak-style": [
            2,
            "unix"
        ],
        "semi": [
            1,
            "never"
        ],
        "no-extra-semi": [
            1
        ],
        "id-length": [
            1,
            {
                min : 1,
                max : 15,
                properties: "never"
            }
        ]
    }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
'use strict'

module.exports = async (args) => {
    // ...
}
```

```bash
eslint test.js
```

**What did you expect to happen?**

space after `async` keyword is allowed, as these are synonymous:

```
fn1 = async function(args) {}
fn2 = async (args) => {}
fn3 = async args => {}
```

**What actually happened? Please include the actual, raw output from ESLint.**

space after `async` keyword is not allowed.

>  3:23  warning  Unexpected space before function parentheses  space-before-function-paren
>
> ✖ 1 problem (0 errors, 1 warning)
>  0 errors, 1 warning potentially fixable with the `--fix` option.

**Comment by not-an-aardvark** [comment]
Hi,

It seems like the rule is currently configured to disallow spaces before function parentheses. Have you tried removing the space?

```js
'use strict'

module.exports = async(args) => {
    // ...
}
```

Alternatively, if you'd like to always have spaces after the `async` keyword for arrow functions, you can configure the rule differently for async arrow functions by adding this to your config:

```json
{
  "rules": {
    "space-before-function-paren": ["error", {
      "anonymous": "never",
      "named": "never",
      "asyncArrow": "always"
    }]
  }
}
```

**Comment by Redsandro** [comment]
Hi @not-an-aardvark,

I think I misunderstood the function of this rule. I thought it removes the space after the `function` keyword, not other keywords.

Thank you for the advanced configuration example.

**Comment by platinumazure** [comment]
Hi @Redsandro, the [`keyword-spacing`](https://eslint.org/docs/rules/keyword-spacing) rule is the rule that enforces spacing around the `function` keyword. Hope this helps!
